staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -711,13 +711,23 @@ func (w *watchCache) getAllEventsSinceLocked(resourceVersion uint64) (*watchCach
 	}
 
 	if resourceVersion == 0 {
-		// resourceVersion = 0 means that we don't require any specific starting point
-		// and we would like to start watching from ~now.
-		// However, to keep backward compatibility, we additionally need to return the
-		// current state and only then start watching from that point.
-		//
-		// TODO: In v2 api, we should stop returning the current state - #13969.
-		return w.getIntervalFromStoreLocked()
+		if opts.SendInitialEvents == nil {
+			// resourceVersion = 0 means that we don't require any specific starting point
+			// and we would like to start watching from ~now.
+			// However, to keep backward compatibility, we additionally need to return the
+			// current state and only then start watching from that point.
+			//
+			// TODO: In v2 api, we should stop returning the current state - #13969.
+			return w.getIntervalFromStoreLocked()
+		}
+		if *opts.SendInitialEvents {
+			// this is a programmer error
+			return nil, errors.NewInternalError(fmt.Errorf("opts.SendInitialEvents cannot be set to true"))
+		}
+		// SendInitialEvents = false and resourceVersion = 0
+		// means that the request would like to start watching
+		// from Any resourceVersion
+		resourceVersion = w.resourceVersion
```

Comment:
I decided to rework the `getAllEventsSinceLocked` function because I would like to avoid acquiring the lock in the cacher

Reply:
if this is okay then i will start fixing failing unit tests.

Reply:
Because without this code we would essentially send `InitialEvents` for `RV=0,sendInitialEvents=false` case.

Reply:
I still don't follow - why the tests are passing then?

Reply:
OK - so I think that with the addition in line 695 - we handle sendInitialEvents=true, above we handle sendInitialEvents=nil, so we're fine now.

Regarding returning RV from this function - I don't think it's needed:
- if RV != 0 => then everything works fine [as we don't change rv here]
- if RV = 0 => we can start the watch from any single point in time, so the only risk we have is that we return some event more than one or we mess up the order of events (or both) - however
(a) in interval processing there is no way it can happen as we process events in the order
(b) there is a risk that we will be processing callbacks for some suffix of the events again (watchcache was updated but processing didn't yet fire), but for this case this if inside watch processing:
https://github.com/kubernetes/kubernetes/blob/ecaf2093f51fed5f544520b0ac00fb33a474b7f5/staging/src/k8s.io/apiserver/pkg/storage/cacher/cache_watcher.go#L492-L494
together with passing the result of that here:
https://github.com/kubernetes/kubernetes/blob/ecaf2093f51fed5f544520b0ac00fb33a474b7f5/staging/src/k8s.io/apiserver/pkg/storage/cacher/cache_watcher.go#L506

is protecting us from that usecase

So we're fine here - no need to return RV from here.